receiver/awscontainerinsightreceiver/internal/efa/efaSysfs.go

```diff
@@ -0,0 +1,443 @@
+		}
+	}
+
+	return errs
+}
+
+func defaultSysFsReader(logger *zap.Logger) sysFsReader {
+	return &sysfsReaderImpl{
+		logger: logger,
+	}
+}
+
+type sysfsReaderImpl struct {
+	logger *zap.Logger
+}
+
+var _ sysFsReader = (*sysfsReaderImpl)(nil)
+
+func (r *sysfsReaderImpl) EfaDataExists() (bool, error) {
+	info, err := os.Stat(efaPath)
+	if err != nil {
+		if os.IsNotExist(err) {
+			return false, nil
+		}
+		return false, err
+	}
+
+	stat, ok := info.Sys().(*syscall.Stat_t)
+	if !ok {
+		return true, nil
+	}
+
+	fmt.Printf("EFA directory uid: %d, perms: %s\n", stat.Uid, info.Mode().Perm())
+
+	if stat.Uid != 0 {
+		r.logger.Warn("EFA directory exists but is not owned by root, not reading from it", zap.String("path", efaPath), zap.Uint32("owner uid", stat.Uid))
+		return false, nil
+	}
+	perms := info.Mode().Perm()
+	if perms&0002 != 0 {
+		r.logger.Warn("EFA directory exists but is writeable by anyone, not reading from it", zap.String("path", efaPath), zap.String("permissions", perms.String()))
+		return false, nil
+	}
+
+	return true, nil
+}
+
+func (r *sysfsReaderImpl) ListDevices() ([]efaDeviceName, error) {
+	dirs, err := os.ReadDir(efaPath)
+	if err != nil {
+		return nil, fmt.Errorf("failed to list EFA devices at %q: %w", efaPath, err)
+	}
+
+	result := make([]efaDeviceName, 0)
+	for _, dir := range dirs {
+		if !dir.IsDir() && dir.Type()&os.ModeSymlink == 0 {
+			continue
+		}
+		result = append(result, efaDeviceName(dir.Name()))
+	}
+
+	return result, nil
+}
+
+func (r *sysfsReaderImpl) ListPorts(deviceName efaDeviceName) ([]string, error) {
+	portsPath := filepath.Join(efaPath, string(deviceName), "ports")
+	portDirs, err := os.ReadDir(portsPath)
+	if err != nil {
+		return nil, fmt.Errorf("failed to list EFA ports at %q: %w", portsPath, err)
+	}
+
+	result := make([]string, 0)
+	for _, dir := range portDirs {
+		if !dir.IsDir() {
+			continue
+		}
+		result = append(result, dir.Name())
+	}
+
+	return result, nil
+}
+
+func (r *sysfsReaderImpl) ReadCounter(deviceName efaDeviceName, port string, counter string) (uint64, error) {
+	path := filepath.Join(efaPath, string(deviceName), "ports", port, "hw_counters", counter)
+	return readUint64ValueFromFile(path)
+}
+
+func readUint64ValueFromFile(path string) (uint64, error) {
+	bytes, err := os.ReadFile(path)
+	if err != nil {
+		if os.IsNotExist(err) || os.IsPermission(err) || err.Error() == "operation not supported" || err.Error() == "invalid argument" {
+			return 0, nil
+		}
+		return 0, fmt.Errorf("failed to read file %q: %w", path, err)
+	}
+	stringValue := strings.TrimSpace(string(bytes))
+
+	// Ugly workaround for handling https://github.com/prometheus/node_exporter/issues/966
+	// when counters are `N/A (not available)`.
+	// This was already patched and submitted, see
+	// https://www.spinics.net/lists/linux-rdma/msg68596.html
+	// Remove this as soon as the fix lands in the enterprise distros.
+	if strings.Contains(stringValue, "N/A (no PMA)") {
```

Comment:
will this break if the fix gets promoted? how do we track this change promotion?

Reply:
Good question.  I'm going to follow up with a commit to improve partial error handling - still post node-level metrics if we couldn't get pod resources data, as we discussed in the meeting; and also include data for some devices/counters even if reading some other data fails.  As part of that commit, I will remove this logic and just skip if parsing to int failed.